105 explore address matching quality

crispy-wonton opened the pull request:
Addresses #105

Description
Exploratory analysis to assess quality of address matching between MCS and EPC addresses, explore causes of errors, and identify potential optimisation of address score threshold. Files:
1. `get_address_matching_sample.py` 
- takes sample of 1002 non-identical matched addresses for manual review
2. `explore_address_matching_sample.py`
- uses manually reviewed address_matching sample to calculate error rate of address matching and overall error rates in population as well as confidence intervals
- explores patterns in errors and causes of errors
- explores address score distribution across datasets
- explores potential for address score threshold optimisation
3. `requirements.txt` 
- contains additional packages required for running this analysis

Set up:
`explore_address_matching_sample.py`
- activate your core conda environment and then pip install requirements to install additional libraries used
- convert `explore_address_matching_sample.py` to `.ipynb` file and run to see the output tables and visualisations
- run the notebook - it should not require any manual inputs/changes to run

`get_address_matching_sample.py`
- you will need to add your path to local ASF core data to the variable indicated at the beginning of the script, other than that it should work without making any changes.
- please can you run this to check that it works - I think it should work as a .py script but I haven't tested it out
- Note this script takes a while (~10-15 mins) to run because of loading and joining large datasets.
- you do not have to run this before `explore_address_matching_sample.py` as the exploratory notebook pulls all files it uses from S3.

Instructions for reviewer:
- please could you check through the code and check that it all runs ok, and also that the general methodology is appropriate

Please pay special attention to:
- sampling methodology and manual review methodology - do these make sense and are there any changes you would make to make the analysis more robust?
- confidence interval method - I used Wilson based on some reading but definitely not an expert on which method is best so appreciate any inputs on selecting the best one for this purpose. I tested all the methods that statsmodels offers and they all seem to produce roughly the same confidence interval range +/- 1 on either side.
- optimising address score threshold - what's your opinion on the findings?


Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

crispy-wonton committed:
update requirements with packages for exploratory address matching analysis

crispy-wonton committed:
add script to generate address matching sample for manual review

crispy-wonton committed:
add script to explore address matching
script investigates error rate of currently implemented address matching as well as patterns in errors. It also explores potential address score threshold optimisation.

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
To transform into notebook I did:

- ` pip install jupytext`
- `jupytext --to notebook asf_core_data/notebooks/explore_address_matching_sample.py`

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# As there is only 1 address at the postcode containing 'example farm', we assume the match is correct and the pairing is given a confident_match of 1.
+#
+# #### Caveats
+# - For ease of reading, we refer in the following analysis to address matches with a confident_match score of 0 as errors. As you can see from the methodology, it is possible that some of these cases are actually true matches but we do not have the data to confidently confirm it. It is therefore possible that this methodology overestimates false positives.
```
When you say the methodology overestimates false positives, you mean your own methodology for reviewing right?

crispy-wonton replied:
Yes

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# %%
+## Params
+
+## Set your sample params
+date = "20231218"
+country_sample_size = 334
+seed = 7
```
might be worth adding a bit of info on what these mean

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
it's perhaps worth adding a README explaining that the two notebooks are related and high level information about them - in case we add more notebooks to this folder in the future

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# After manually checking through the sampled rows of matching addresses, we will assess the accuracy of address matching for each country.
+
+# %%
+mcs_epc = pd.read_csv(
+    f"s3://asf-heat-pump-suitability/address-matching/{date}_mcs_epc_address_join_23Q2_prepr_dedupl_231009.csv",
+    keep_default_na=False,
+)
+sample = pd.read_csv(
+    f"s3://asf-heat-pump-suitability/address-matching/{date}_mcs_epc_address_matching_sample_n{country_sample_size*3}_seed{seed}_reviewed.csv",
+    index_col=0,
+)
```
I'm also missing some info on what these two datasets represent.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
Had to do `export PYTHONPATH=$PWD` before running this as as script, i.e. `python asf_core_data/notebooks/get_address_matching_sample.py`


I also think we need the `os.makedirs()` I added below, as the path does not exist, but otherwise ran fine.

crispy-wonton replied:
Thanks, good point!

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+local_core_data_dir = ""
+
+## Output data will be saved here
+outputs_data_dir = os.path.abspath("../../outputs/data")
```
If you do it in the following way, the output will always be saved in the same place, independently of where you're running it from.

And because the folder data doesn't exist, we need to create it.
```suggestion
outputs_data_dir = os.path.join(PROJECT_DIR,"outputs/data")
os.makedirs(outputs_data_dir, exist_ok=True)
```

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# MCS data: mcs_installations_231009
+
+# %%
+epc_data = pd.read_csv(
+    os.path.join(
+        local_core_data_dir,
+        "outputs/EPC/preprocessed_data/2023_Q2_complete/EPC_GB_preprocessed_and_deduplicated.csv",
+    )
+)
+mcs_data = pd.read_csv("s3://asf-core-data/outputs/MCS/mcs_installations_231009.csv")
```
Absolutely no need to change this, but for future reference, you can use the existing getters, e.g. :
```
from asf_core_data.getters import data_getters
from asf_core_data import load_preprocessed_epc_data

data_getters.download_core_data('epc_preprocessed_dedupl', local_core_data_dir, batch="231009")
epc_data = load_preprocessed_epc_data(data_path=local_core_data_dir, version='preprocessed_dedupl')
```

You could even check if the data is stored locally to only download it once. Also to ensure that the version you need is actually stored locally.

crispy-wonton replied:
Oh yeah! Thanks for the tip!

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# MCS data: mcs_installations_231009
+
+# %%
+epc_data = pd.read_csv(
```
something worth checking:
- I think (?) the matching is done with the non deduplicated version and this might change results slightly in case the same UPRN has different addresses (which I imagine it might happen due to UPRN errors);

crispy-wonton replied:
Yes I think you are correct because [line 139-144 in `pipeline/mcs/generate_mcs_data.py` ](https://github.com/nestauk/asf_core_data/blob/15ccd754dbc31c2a8ba4d84463cba7ebfe5ebf82/asf_core_data/pipeline/mcs/generate_mcs_data.py#L139) uses default `epcs` parameter, which is `None`, and sets `all_records=True`. Lines [293-294 in `pipleine/mcs/process/mcs_epc_joining.py`](https://github.com/nestauk/asf_core_data/blob/15ccd754dbc31c2a8ba4d84463cba7ebfe5ebf82/asf_core_data/pipeline/mcs/process/mcs_epc_joining.py#L293) show this would set the `epc_version` to `"preprocessed"`.

I will correct this.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# ## Prepare data for address matching
+
+# %%
+epc_df = joining.prepare_epcs(epc_data)
+
+# %%
+mcs_df = joining.prepare_hps(mcs_data)
```
I think you told me that `prepare_epcs` and `prepare_hps` are doing slightly different things. Will this have an impact on results?

crispy-wonton replied:
Yes, `prepare_hps` concatenates `address_1` and `address_2` from mcs address fields and then standardises them with by removing punctuation and converting to lower case. `prepare_epcs` only standardises `ADDRESS1` field without concatenating any other address lines. I've now updated the `prepare_epcs` function to use `ADDRESS2` and will review the effect. I think it will have some impact but I think a lot of the initial errors discovered will still be present.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# Here we calculate an address match score for all candidate matches (i.e. all addresses in the same postcode).
+
+# %%
+matches = joining.form_matching(mcs_df, epc_df)
```
I am assuming you're following the exact same methodology as followed in the joining script? - I know you probably couldn't call the exact function used to do the matching because you would miss some variables you need to evaluate the match, but just worth double checking that the methodology followed is exactly the same.

The reason why I'm saying this is that the match is formed in this [script](https://github.com/nestauk/asf_core_data/blob/15ccd754dbc31c2a8ba4d84463cba7ebfe5ebf82/asf_core_data/pipeline/mcs/generate_mcs_data.py) when calling the `join_mcs_epc_data()` function, which uses the `form_matching()` function above.

sofiapinto replied:
Understood after going through the whole script :)

crispy-wonton replied:
Yes it should be exactly the same and the output of both joining methodologies is tested below to ensure that they give the same output :)

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# %% [markdown]
+# ## Join walkthrough
+#
+# Here we will join EPC data to MCS data via address matching using the function used in the asf_core_data pipeline. The function is modified here to retain standardised joining cols and address score. All top matches of EPC to MCS (equivalent to setting all_records=True) will be kept.
```
I think this answers my question above, but i'll keep it there just in case :)

sofiapinto replied:
What do you mean by "All top matches of EPC to MCS (equivalent to setting all_records=True) will be kept."?

crispy-wonton replied:
Yes exactly, this is where we test that both methodologies give the same output - as we would expect if the code in the notebook is written correctly.

The 'all top matches' is terminology from [lines 161-163 in the `join_prepared_mcs_epc_data` function](https://github.com/nestauk/asf_core_data/blob/c7886c041ed08d352a4ec6e088c1e26df90d549c/asf_core_data/pipeline/mcs/process/mcs_epc_joining.py#L161). I believe it means that all matched rows for a given address which have the same `address_score` will be kept (i.e. multiple records for the same property). If you set `all_records=False`, you will just get the first occurrence of a given match.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# %%
+# Keep standardised address, postcode and numeric tokens columns for reference later
+
+hps = mcs_df.copy()
+epcs = epc_df.copy()
```
why do we need to do this?

crispy-wonton replied:
It's not necessary in the pipeline but I did it in the notebook to keep those variables so I didnt have to regenerate them every time I made changes as they take a while.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+
+# %%
+# Join the index-matching df on MCS index to the MCS records
+merged = hps.merge(top_matches, how="left", left_on="index", right_on="level_0")
```
I know this probably comes from the original methodology - but are we sure "index" and "level_0" are the same thing? Taking a quick look, it doesn't look like so.

crispy-wonton replied:
Yes I think `level_0` contains numbers representing indices in `hps.index` and `level_1` contains numbers representing indices in `epcs.index`. I've extracted a couple of example indices and checked that they correlate to a sensible row in `epcs` and `hps` and it checks out. I also checked that all values in `level_0` are found in `hps.index` and same with `level_1` and `epcs`.

Is there something else that made you think they were different?

sofiapinto replied:
it's just that I don't see where `level_1`/ `level_0`are coming from - maybe they're computed prior to this?

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+
+# %%
+# See matching with existing address-matching function
+mcs_epc_join_function = joining.join_prepared_mcs_epc_data(
```
should it be `join_mcs_epc_data()` as this is the one being used [here](https://github.com/nestauk/asf_core_data/blob/15ccd754dbc31c2a8ba4d84463cba7ebfe5ebf82/asf_core_data/pipeline/mcs/generate_mcs_data.py)?

crispy-wonton replied:
Yes good point, thank you! I think it should give the same output with either function as `join_mcs_epc_data()` calls `join_prepared_mcs_epc_data()` to do the actual join and then the function also includes the preprocessing steps. But I will update it with `join_mcs_epc_data()` to confirm that the output is the same.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+)
+
+# %%
+_merged = merged.drop(
+    columns=[
+        "mcs_address",
+        "mcs_postcode",
+        "mcs_numeric_tokens",
+        "epc_postcode",
+        "epc_numeric_tokens",
+        "address_score",
+    ]
+)
+assert mcs_epc_join_function.equals(_merged)
```
amazing! ⭐️ I see you're checking that you're getting the same thing - and so you answered my question above! I'm really just not sure if using `join_mcs_epc_data()` and `join_prepared_mcs_epc_data()` is the same.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+)
+
+# %%
+## Remove rows with no EPC match
+## Any rows with all NaN values across all EPC columns in mcs_epc df will be removed
+epc_cols = epc_df.columns.to_list()
+epc_cols.extend(["epc_address", "epc_postcode", "epc_numeric_tokens"])
```
what about removing EPC instances with no match to MCS? Or are those already not in this dataframe?

crispy-wonton replied:
I think those don't exist in the dataset because everything is left-joined onto the MCS data.

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
```diff
@@ -0,0 +1,318 @@
+# %%
+## We will take 334 rows from each country to create a sample of 1002 address matches
+country_sample_size = 334
+seed = 7
```
you selected this one just so you could get the same sample every time you run this script, am I right?

crispy-wonton replied:
Yes correct :)

sofiapinto commented on asf_core_data/notebooks/get_address_matching_sample.py:
The methodology looks good! I left a couple of questions, but they're mostly about the matching, not the sampling itself.

In terms of the sampling, there are obviously multiple ways of going about this. Doing it this way, the only thing we're sure is that we have the same number of instances from each country. The sampling does not account for the type of home or other variables, so we might get a highest proportion of a specific type of home and very low of another - but that's the only thing I would point out. I guess you can also check the numbers _a posteriori_ and see if you're happy w/ them.

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# ---
+
+# %% [markdown]
+# ### Methodology for manually reviewing address matching sample
```
Methodology looks good! I guess the only thing I would have done differently would be to tag the "not sure's" as -1 just so you differentiate them from the confident 0s. 

I guess in a way it is important to identify the "not sure's" to answer: is the model doing poorly or is it just doing badly when the data isn't good enough?

crispy-wonton replied:
Great point for important distinction. And I guess if we have that info we can always change it to a binary score later if necessary but good to differentiate.

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+    .value_counts()
+    .reset_index(name="counts")
+)
+sample_country_split.groupby("country")["counts"].sum()
```
either print, or remove I guess :)

```suggestion
```

crispy-wonton replied:
Great spot thank you

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+
+# %%
+## Calculate confidence intervals for error rates at low and higher bounds in the sample
+## Wilson interval selected as it's recommended for real data applications
```
I don't have a lot of experience with these methods, but after a quick read this one seems to be good for small samples, so sounds good!

crispy-wonton replied:
Great, thanks for looking that up!

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+    .size()
+    .reset_index(name="non_exact_match_counts_popn")
+)
+country_totals = country_totals[country_totals["exact_match"] != 1]
+country_totals = country_totals.drop(columns="exact_match")
```
so basicaly we're computing the non exact matches for the whole dataset, so that we can then used the estimated proportion wrongly matched to compute how many wrongly matched we will have in the whole dataset, is that it?

crispy-wonton replied:
yep that's the idea :)

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+)
+
+## Get rate of errors across conf int for each country
+popn_error_conf_int["rate_errors_popn"] = (
+    popn_error_conf_int["num_errors_popn"] / popn_error_conf_int["total_counts_popn"]
+)
+popn_error_conf_int["rate_lower_errors_popn"] = (
+    popn_error_conf_int["num_lower_errors_popn"]
+    / popn_error_conf_int["total_counts_popn"]
+)
+popn_error_conf_int["rate_upper_errors_popn"] = (
+    popn_error_conf_int["num_upper_errors_popn"]
+    / popn_error_conf_int["total_counts_popn"]
+)
```
sorry if i'm missing something here - but why are these proportions different from the above?

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+    print(len(val))
+
+# %% [markdown]
+# MCS addresses with NaN or punctuation characters only are converted to a space character (due to this line https://github.com/nestauk/asf_core_data/blob/c7886c041ed08d352a4ec6e088c1e26df90d549c/asf_core_data/pipeline/mcs/process/mcs_epc_joining.py#L51). This space character then matches with short epc_addresses (seemingly 10 or fewer characters) containing a space. This can easily be fixed in t ...
```
this is so weird. how can a space end up matched to a actual existing address?

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# - Similar spelling of house names in the same postcode
+
+# %%
+print(
+    f"Number of errors in addresses in sample with numeric components: {len(sample_error_num)}"
+)
+print(
+    f"% of errors in sample with numeric components: {round(100 * len(sample_error_num)/errors, 2)}%"
+)
+print(
+    f"Number of errors in addresses in sample without numeric components: {len(sample_error_alpha)}"
+)
+print(
+    f"% of errors in sample without numeric components: {round(100 * len(sample_error_alpha)/errors, 2)}%"
+)
```
this might not mean that there's a bigger tendency for errors to occur in one other than the other, but that your sample has one more than the other

crispy-wonton replied:
Yes good point - I'll need to update it so it's not misleading

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+
+_sample = sample.copy()
+
+_sample["address_score_qbin20"] = pd.qcut(_sample["address_score"], 20)
```
cool - have never used the `qcut` function before :)

sofiapinto replied:
perhaps a note on what the 20 means might be good :)

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# %%
+## Cut for equal sized ranges
+
+_sample["address_score_bin15"] = pd.cut(_sample["address_score"], 15)
+_match_rate = _sample.groupby("address_score_bin15").agg({"confident_match": np.mean})
+
+fig, ax = plt.subplots(figsize=(10, 5))
+
+sns.barplot(x=_match_rate.index, y=_match_rate["confident_match"], ax=ax)
+
+ax.set_title(f"Confident Match rate by address score on sample, N={len(_sample)}")
+
+plt.xticks(rotation=90)
+
```
little confused. What is the difference between this and the above?

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+)
+
+# %% [markdown]
+# It appears that erroneous matches of addresses with numeric components tend to have lower address scores and erroneous matches of addresses with no numeric components tend to have higher or very low address scores.
```
low count though, hard to take conclusions.

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+rcd.from_predictions(y_true=sample["confident_match"], y_pred=sample["address_score"])
+
+# %%
+youdens_thresholds_country = {}
+for country in ["England", "Scotland", "Wales"]:
+    fpr, tpr, thresholds = roc_curve(
+        y_true=sample[sample["country"] == country]["confident_match"],
+        y_score=sample[sample["country"] == country]["address_score"],
+    )
+    ## Calculate Youden's J statistic
+    ## sensitivity + specificity - 1 = TPR - FPR
+    _df = pd.DataFrame({"youdens": tpr - fpr, "threshold": thresholds}).sort_values(
+        by="youdens", ascending=False
+    )
+    _df["removed_candidates"] = [
+        (sample["address_score"] < threshold).sum() for threshold in _df["threshold"]
+    ]
+    _df["removed_true_matches"] = [
+        (sample[sample["confident_match"] == 1]["address_score"] < threshold).sum()
+        for threshold in _df["threshold"]
+    ]
+    youdens_thresholds_country[country] = _df
+
+youdens_thresholds_country
```
perhaps some documentation on what you're doing here would be helpful for future reference. e.g. why are you looking at the Youden's J statistic and what exactly you are doing. At the moment it's a bit hard to follow.

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+# - We want to maximise Youden's J-statistic
+# - i.e. sensitivity and specificity are equally important: removing false positives is as important as retaining true positives
+#
+# Based on these results and the confident match rate by address scores assessed above, we should consider increasing the address score threshold at least to 0.74 and potentially to somewhere between 0.77 and 0.87.
```
where are these values coming from? you're ouputing the whole df's above, so how are these 0.77 and 0.87 computed?

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+youdens_thresholds_country
+
+# %%
+youdens_thresholds_numeric = {}
+for _, key in zip([0, 1], ["non_numeric", "numeric"]):
+    fpr, tpr, thresholds = roc_curve(
+        y_true=sample[sample["numeric_address_component"] == _]["confident_match"],
+        y_score=sample[sample["numeric_address_component"] == _]["address_score"],
+    )
+    ## Calculate Youden's J statistic
+    ## sensitivity + specificity - 1 = TPR - FPR
+    _df = pd.DataFrame({"youdens": tpr - fpr, "threshold": thresholds}).sort_values(
+        by="youdens", ascending=False
+    )
+    _df["removed_candidates"] = [
+        (sample["address_score"] < threshold).sum() for threshold in _df["threshold"]
+    ]
+    _df["removed_true_matches"] = [
+        (sample[sample["confident_match"] == 1]["address_score"] < threshold).sum()
+        for threshold in _df["threshold"]
+    ]
+    youdens_thresholds_numeric[key] = _df
+
+youdens_thresholds_numeric
```
same as above

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+youdens_thresholds_numeric
+
+# %% [markdown]
+# Even when we search for optimum threshold by country or by numeric/alpha address, it looks like there could be a potential optimum address score threshold between 0.84-0.869231.
```
where are these numbers coming from?

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
I found this one a bit more difficult to follow than the other one, so maybe adding some more documentation would help. The analysis is truly very thorough - thanks so much for that!

sofiapinto commented on asf_core_data/notebooks/explore_address_matching_sample.py:
```diff
@@ -0,0 +1,485 @@
+#
+# Based on these results and the confident match rate by address scores assessed above, we should consider increasing the address score threshold at least to 0.74 and potentially to somewhere between 0.77 and 0.87.
+#
+# However, we lose true positive matches with every increase in the threshold. Also, simply increasing the threshold cannot remove many of the erroneous matches in non-numeric addresses because many of these have very high similarity scores. We will need a different methodology to address these errors. We will explore alternative improvements to address matching in a separate analysis.
```
I guess a question I have is: is the best we can get from this model (by doing the above improvements) enough? it's certainly better than we have now, but we potentially need a better address matching model.

sofiapinto reviewed (commented):
hey there @crispy-wonton, great work as always!

- very thorough analysis of the errors, although sometimes i found it a bit hard to follow (i think it just needs a bit more documentation)
- I ran the two notebooks/scripts - all good - left 1 or 2 notes about this;
- I mostly left comments/questions for clarification!
- I've also left comments about the sampling methodology, the interval method (although, I didn't have much to add) and about optimising the address score threshold.

crispy-wonton committed:
fix errors in original address matching pipeline
- join address lines 1 and 2 of EPC
- fix bug that generated additional white space characters in mcs_address field creation

crispy-wonton committed:
add AddressParser preprocessing functions
